website/blog/2024/google-search-crawler-draft/index.md

```diff
@@ -0,0 +1,201 @@
+
+If Google search isn't going anywhere in the coming years, then tasks like [`SERP Analysis`](https://www.semrush.com/blog/serp-analysis/), SEO optimization, and evaluating your site's search ranking remain highly relevant - and for these, we need effective tools.
+
+That's why in this blog, we'll explore creating a Google search results crawler with [`Crawlee for Python`](https://crawlee.dev/python). This will be particularly useful if you're conducting a small data analysis project, analyzing search results for your business, or writing an [article about Google ranking analysis](https://backlinko.com/search-engine-ranking). Rather ironically, don't you think, considering that the largest known crawler belongs to Google itself?
+
+Let's get started!
+
+## 1. Site Analysis
+
+First, let's define the scope of what we want to extract. Google's search results are quite heterogeneous now, featuring map widgets, notable people, company information, video blocks, popular questions, and much more. Let's limit ourselves to analyzing classic search results with rankings.
+
+Here's what we'll be extracting:
+
+![Search Example](./img/search_example.webp)
+
+Let's verify whether we can extract the necessary data from the page's HTML code or if we need deeper analysis or `JS` rendering. Note that this verification is sensitive to HTML tags:
+
+![Check Html](./img/check_html.webp)
+
+Based on the data obtained from the page, all necessary information is present in the HTML code. Therefore, we can use [`beautifulsoup_crawler`](https://crawlee.dev/python/docs/examples/beautifulsoup-crawler).
+
+Let's analyze the HTML code of the elements we need to extract:
+
+![Check Html](./img/html_example.webp)
+
+### Important Considerations
+
+#### 1. Tag Attributes
+
+There's an obvious distinction between *readable* id attributes and *generated* class names and other attributes. When creating selectors for data extraction, you should ignore any generated attributes. Even if you've read that Google has been using a particular generated tag for N years, you shouldn't rely on it - this is your experience in writing robust code.
+
+#### 2. Complex Selectors
+
+The abundance of generated tags and Google's data structure can be overwhelming. Don't hesitate to use advanced features within CSS and XPath selector syntax.
+
+#### 3. Write Selectors Manually
+
+For any project that needs to be reproduced more than once or any large-scale website project that might have multiple page layouts, you should avoid using selectors obtained from browser's "Copy XPath/selector" feature. Selectors built with absolute paths are not stable against any page changes. You should write selectors yourself, so familiarize yourself with the basics of [CSS](https://www.w3schools.com/cssref/css_selectors.php) and [XPath](https://www.w3schools.com/xml/xpath_syntax.asp) syntax.
```

Comment:
Here are 2 selectors obtained from dev tool one for `google` the other for `g2.com`
//*[@id="rso"]/div[2]/div[8]/div/div/div/div[2]/div
/html/body/div[5]/div/div/div[1]/div/div[5]/div/div[2]/div[5]/div[1]

The first one looks good, at least because it uses @id, but if tomorrow google updates the page markup by making at least one div smaller, the selector won't work or will return wrong data.

For g2.com it just looks terrible.

These selectors work well for some quick tests here and now, but no more.